bs_1/src/telemetry.cpp

```diff
@@ -0,0 +1,243 @@
+
+#include "target.h"
+#include "ser_des.h"
+
+#ifdef TELEMETRY_BASE_STATION_TX
+  // CAN library for Teensy
+  #include "teensy_can.h"
+#endif
+
+/********** DEFINES **********/
+
+/* Packet size */
+// Size of data packet to be sent over LoRa
+// Math is conducted below:
+#define PACKET_SIZE 23
+
+/********** VARIABLES **********/
+
+/* RadioHead */
+// Driver
+RH_RF95 rf95(RFM95_CS, RFM95_INT);
+
+// Packet number for ordering/debugging
+int16_t packetnum = 0;
+
+// Success
+bool rfm95_init_successful = true;
+
+#ifdef TELEMETRY_BASE_STATION_TX
+  // Initialize bus
+  TeensyCAN<1> can_bus{};
+
+  /* CAN data buffers */ 
+  // Each signal is 16-bit with 10 sigs in total
+  // Total data: 160 bits = 20 bytes (chars)
+  CANSignal<float, 0, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(0)> fl_wheel_speed_sig;
+  CANSignal<float, 0, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(0)> fr_wheel_speed_sig;
+  CANSignal<float, 0, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(0)> bl_wheel_speed_sig;
+  CANSignal<float, 0, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(0)> br_wheel_speed_sig;
+
+  CANSignal<float, 16, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(-40)> fl_brake_temperature_sig;
+  CANSignal<float, 16, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(-40)> fr_brake_temperature_sig;
+  CANSignal<float, 16, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(-40)> bl_brake_temperature_sig;
+  CANSignal<float, 16, 16, CANTemplateConvertFloat(0.1), CANTemplateConvertFloat(-40)> br_brake_temperature_sig;
+
+  CANRXMessage<2> fl_wheel_msg{can_bus, 0x400, fl_wheel_speed_sig, fl_brake_temperature_sig};
+  CANRXMessage<2> fr_wheel_msg{can_bus, 0x401, fr_wheel_speed_sig, fr_brake_temperature_sig};
+  CANRXMessage<2> bl_wheel_msg{can_bus, 0x402, bl_wheel_speed_sig, bl_brake_temperature_sig};
+  CANRXMessage<2> br_wheel_msg{can_bus, 0x403, br_wheel_speed_sig, br_brake_temperature_sig};
+
+  CANSignal<uint16_t, 0, 16, CANTemplateConvertFloat(1), CANTemplateConvertFloat(0)> front_brake_pressure_sig;
+  CANSignal<uint16_t, 16, 16, CANTemplateConvertFloat(1), CANTemplateConvertFloat(0)> rear_brake_pressure_sig;
+
+  CANRXMessage<2> brake_pressure_msg{can_bus, 0x410, front_brake_pressure_sig, rear_brake_pressure_sig};
+
+  // Additional 3 bytes appended at end: 2 bytes for packetnum, 1 byte for null-terminator
+  // Total packet size: 23 bytes < capacity
+
+  /* Packet */
+  // Common across send and receive functions
+  char packet[PACKET_SIZE];
+#endif
+
+#ifdef TELEMETRY_BASE_STATION_TX
+  // Raw signal data
+  uint16_t fl_wheel_speed;
+  uint16_t fl_brake_temperature;
+  uint16_t fr_wheel_speed;
+  uint16_t fr_brake_temperature;
+  uint16_t bl_wheel_speed;
+  uint16_t bl_brake_temperature;
+  uint16_t br_wheel_speed;
+  uint16_t br_brake_temperature;
+
+  uint16_t front_brake_pressure;
+  uint16_t rear_brake_pressure;
+#endif
+
+// #ifdef TELEMETRY_BASE_STATION_RX
```

Comment:
Why are these commented out? Remove if not used

Reply:
bs_template is used to provide template code for subsequent projects, to automate copy-pasting. bs_1 is not the PDR project; lora_can_2 is. The commit "Final program for PDR" is the one I wanted to make the PR for; does Github not keep track of when the PR is made in between commits?